File: packages/core/src/executors/http-executor.ts
import axios, { AxiosRequestConfig } from 'axios';
import { ToolDefinition } from '../core/schema';
import { applyParameterEncodings } from '../encodings/parameter-encoding';
import { MatimoError, ErrorCode, fromHttpError } from '../errors/matimo-error';

/**
 * HttpExecutor - Executes HTTP requests
 * Handles authentication, retries, and response validation
 */

export class HttpExecutor {
  /**
   * Execute a tool that makes an HTTP request.
   *
   * @param tool - Tool definition
   * @param params - Tool parameters (already env-injected by MatimoInstance)
   * @param credentials - Optional per-call credential overrides. Used for
   *   `authentication.type: basic` (username_env / password_env keys) instead of
   *   reading from `process.env`. Other auth schemes (bearer, api_key) are handled
   *   upstream via parameter templating in MatimoInstance.injectAuthParameters().
   *   Values are never logged.
   */
  async execute(
    tool: ToolDefinition,
    params: Record<string, unknown>,
    credentials?: Record<string, string>
  ): Promise<unknown> {
    if (tool.execution.type !== 'http') {
      throw new MatimoError('Tool execution type is not http', ErrorCode.EXECUTION_FAILED, {
        expectedType: 'http',
        actualType: tool.execution.type,
      });
    }

    const {
      method,
      url,
      headers = {},
      body,
      timeout,
      query_params,
      parameter_encoding,
    } = tool.execution;
    const queryParams = query_params;
    const parameterEncodings = parameter_encoding;

    // Apply parameter encodings (e.g., MIME encoding for email parameters)
    let finalParams = params;
    if (parameterEncodings && parameterEncodings.length > 0) {
      finalParams = applyParameterEncodings(params, parameterEncodings);
    }

    // Validate URL parameters are provided
    this.validateUrlParameters(url, finalParams);

    // Implement parameter templating
    let finalUrl = this.templateString(url, finalParams);

    // Handle query parameters - only include non-empty ones
    if (queryParams) {
      const queryString = this.buildQueryString(queryParams, finalParams);
      if (queryString) {
        finalUrl += '?' + queryString;
      }
    }

    const templatedHeaders = this.templateObject(headers, finalParams, tool.parameters);

    // Natively handle HTTP Basic Auth when authentication.type === 'basic' and
    // username_env/password_env are declared. This eliminates the need for
    // developers to pre-compute base64 credentials as a separate env var step.
    // Credentials override takes precedence over process.env when provided.
    this.applyBasicAuth(tool, templatedHeaders as Record<string, string>, credentials);
    const templatedBody =
      body && typeof body === 'object'
        ? this.templateObject(body as Record<string, unknown>, finalParams, tool.parameters)
        : typeof body === 'string'
          ? this.templateString(body, finalParams)
          : body;

    // Build request config
    const requestConfig: AxiosRequestConfig = {
      method,
      url: finalUrl,
    };

    if (Object.keys(templatedHeaders).length > 0) {
      requestConfig.headers = templatedHeaders as Record<string, string>;
    }

    if (templatedBody !== undefined) {
      // If Content-Type is application/x-www-form-urlencoded and body is an object,
      // convert to URLSearchParams for proper form encoding by axios
      const contentType =
        (templatedHeaders as Record<string, string>)['Content-Type'] ||
        (templatedHeaders as Record<string, string>)['content-type'];

      if (
        contentType?.includes('application/x-www-form-urlencoded') &&
        typeof templatedBody === 'object' &&
        templatedBody !== null &&
        !(templatedBody instanceof FormData)
      ) {
        const params = new URLSearchParams();
        for (const [key, value] of Object.entries(templatedBody as Record<string, unknown>)) {
          if (value !== undefined && value !== null) {
            params.append(key, String(value));
          }
        }
        requestConfig.data = params;
      } else {
        requestConfig.data = templatedBody;
      }
    }

    if (timeout !== undefined) {
      requestConfig.timeout = timeout;
    }

    try {
      const response = await axios.request(requestConfig);

      const success = response.status >= 200 && response.status < 300;
      return {
        success,
        data: response.data,
        statusCode: response.status,
        headers: response.headers,
      };
    } catch (error) {
      // If this is already a MatimoError, rethrow to preserve semantics
      if (error instanceof MatimoError) {
        throw error;
      }

      // Normalize any HTTP/axios-like or generic errors into MatimoError.
      // This keeps callers consistent and preserves original cause via details.cause.
      throw fromHttpError(error, 'HTTP request failed');
    }
  }

  /**
   * Automatically inject `Authorization: Basic <base64(username:password)>` when
   * the tool declares `authentication.type: basic` with `username_env` and `password_env`.
   *
   * This is a zero-friction pattern: developers only set two natural env vars
   * (e.g. TWILIO_ACCOUNT_SID + TWILIO_AUTH_TOKEN) and Matimo handles encoding.
   * No pre-computed base64 credential string required.
   *
   * When `credentials` is provided the lookup order is:
   *   1. `credentials[envVarName]`  (per-call override — multi-tenant use)
   *   2. `process.env[envVarName]`  (singleton / single-tenant fallback)
   *
   * Credential values are never logged or included in error details.
   */
  private applyBasicAuth(
    tool: ToolDefinition,
    headers: Record<string, string>,
    credentials?: Record<string, string>
  ): void {
    const auth = tool.authentication;
    if (auth?.type !== 'basic' || !auth.username_env || !auth.password_env) {
      return;
    }

    // Prefer per-call credentials, fall back to process.env
    const username =
      (credentials && credentials[auth.username_env]) ?? process.env[auth.username_env];
    const password =
      (credentials && credentials[auth.password_env]) ?? process.env[auth.password_env];

    if (!username || !password) {
      throw new MatimoError(
        `Basic Auth requires env vars '${auth.username_env}' and '${auth.password_env}' to be set`,
        ErrorCode.AUTH_FAILED,
        { toolName: tool.name, username_env: auth.username_env, password_env: auth.password_env }
      );
    }

    const encoded = Buffer.from(`${username}:${password}`).toString('base64');
    headers['Authorization'] = `Basic ${encoded}`;
  }

  /**
   * Replace parameter placeholders in a string
   */
  private templateString(str: string, params: Record<string, unknown>): string {
    let result = str;
    for (const [key, value] of Object.entries(params)) {
      const placeholder = `{${key}}`;
      // Allow empty-string values to be substituted (validation allows empty strings)
      if (value !== undefined && value !== null) {
        result = result.replace(new RegExp(placeholder, 'g'), String(value));
      }
    }
    return result;
  }

  /**
   * Check if a string is an unfilled placeholder
   * Only matches single placeholders like "{param}", not "{...}" or embedded placeholders
   */
  private isUnfilledPlaceholder(str: string): boolean {
    // Match exactly "{word}" where word is a valid identifier
    return /^\{[a-zA-Z_][a-zA-Z0-9_]*\}$/.test(str);
  }

  /**
   * Validate that all URL parameters are provided
   */
  private validateUrlParameters(url: string, params: Record<string, unknown>): void {
    // Extract all placeholders from URL like {param_name}
    const urlParamMatches = url.match(/\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g);
    if (!urlParamMatches) {
      return; // No parameters in URL
    }

    for (const match of urlParamMatches) {
      const paramName = match.slice(1, -1); // Remove { }
      const paramValue = params[paramName];

      // Check if parameter is missing or undefined
      if (paramValue === undefined) {
        throw new MatimoError(
          `Required URL parameter '${paramName}' is missing`,
          ErrorCode.INVALID_SCHEMA,
          { url, missingParam: paramName }
        );
      }

      // Allow null, empty string, 0, false, etc. as valid values
      // Just ensure the parameter is defined
    }
  }

  /**
   * Build query string from query_params, only including provided values
   */
  private buildQueryString(
    queryParams: Record<string, string>,
    params: Record<string, unknown>
  ): string {
    const parts: string[] = [];
    for (const [key, template] of Object.entries(queryParams)) {
      const value = this.templateString(template, params);
      // Only include if not empty and didn't result in "{param}" (unfilled placeholder)
      if (value && !value.startsWith('{')) {
        parts.push(`${encodeURIComponent(key)}=${encodeURIComponent(value)}`);
      }
    }
    return parts.join('&');
  }

  /**
   * Replace parameter placeholders in an object (headers, body, query params)
   *
   * CORE PRINCIPLE: "Define once in YAML, embed correctly at execution time"
   *
   * This method intelligently handles different parameter types:
   * - STRING placeholders like "{title}": Always templated as strings
   * - OBJECT placeholders like "{parent}": Embedded directly as JSON objects (not stringified) if paramDefinitions specifies type:object
   * - ARRAY placeholders like "{items}": Embedded directly as JSON arrays (not stringified) if paramDefinitions specifies type:array
   *
   * Key behaviors:
   * - Recursively processes nested objects
   * - Skips keys with unfilled placeholders (e.g., "{sort_by}" when sort_by not provided)
   * - Uses parameter schema type from YAML to determine how to embed values
   * - Preserves JSON structure for complex types (objects/arrays) sent to APIs
   *
   * @example
   * ```
   * // YAML definition:
   * parameters:
   *   parent:
   *     type: object  // <-- Tells executor to embed as-is, not stringify
   *   items:
   *     type: array   // <-- Tells executor to embed as-is, not stringify
   *   title:
   *     type: string  // <-- String templating applies
   *
   * body:
   *   parent: "{parent}"  // Object embedded as {"id": "123", ...}
   *   items: "{items}"    // Array embedded as [{"name": "a"}, ...]
   *   title: "{title}"    // String embedded as "My Title"
   *
   * // JavaScript call:
   * const result = await matimo.execute('notion_create_page', {
   *   parent: { database_id: 'abc123' },  // JavaScript object
   *   items: [{ type: 'text' }],          // JavaScript array
   *   title: 'Create This Page'           // String
   * });
   *
   * // HTTP body sent to API:
   * {
   *   "parent": {"database_id": "abc123"},     // Proper JSON object
   *   "items": [{"type": "text"}],            // Proper JSON array
   *   "title": "Create This Page"             // String
   * }
   * ```
   */
  private templateObject(
    obj: Record<string, unknown>,
    params: Record<string, unknown>,
    paramDefinitions?: Record<string, { type: string }>
  ): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj)) {
      if (typeof value === 'string') {
        // Check if this is exactly a placeholder like "{parent}"
        const paramNameMatch = value.match(/^\{([a-zA-Z_][a-zA-Z0-9_]*)\}$/);

        if (paramNameMatch) {
          const paramName = paramNameMatch[1];
          const paramValue = params[paramName];

          // If parameter is defined in schema with type object or array,
          // embed the value directly instead of stringifying
          if (paramValue !== undefined && paramDefinitions?.[paramName]) {
            const paramType = paramDefinitions[paramName].type;
            if ((paramType === 'object' || paramType === 'array') && paramValue !== null) {
              result[key] = paramValue;
              continue;
            }
          }
        }

        // Otherwise, do normal string templating
        const templated = this.templateString(value, params);
        if (
          templated !== undefined &&
          templated !== null &&
          !this.isUnfilledPlaceholder(templated)
        ) {
          // Extract parameter name from template (e.g., "{page_size}" → "page_size")
          const paramName = paramNameMatch ? paramNameMatch[1] : null;

          // If this is a single parameter placeholder, use the parameter's defined type
          if (paramName && paramDefinitions?.[paramName]) {
            const paramType = paramDefinitions[paramName].type;
            if (paramType === 'number' && !isNaN(Number(templated))) {
              result[key] = Number(templated);
            } else if (paramType === 'boolean') {
              result[key] = templated === 'true' || templated === '1' || templated === 'yes';
            } else {
              result[key] = templated;
            }
          } else {
            result[key] = templated;
          }
        }
      } else if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
        // Recursively template nested objects
        const nestedResult = this.templateObject(
          value as Record<string, unknown>,
          params,
          paramDefinitions
        );
        // Only include nested object if it has content
        if (Object.keys(nestedResult).length > 0) {
          result[key] = nestedResult;
        }
      } else if (Array.isArray(value)) {
        // Handle arrays of objects
        const templatedArray = value
          .map((item) => {
            if (typeof item === 'string') {
              const templated = this.templateString(item, params);
              // Skip unfilled placeholders; allow empty strings
              if (
                templated !== undefined &&
                templated !== null &&
                !this.isUnfilledPlaceholder(templated)
              ) {
                // Extract parameter name from template
                const paramNameMatch = item.match(/^\{([a-zA-Z_][a-zA-Z0-9_]*)\}$/);
                const paramName = paramNameMatch ? paramNameMatch[1] : null;

                // If this is a single parameter placeholder, use the parameter's defined type
                if (paramName && paramDefinitions?.[paramName]) {
                  const paramType = paramDefinitions[paramName].type;
                  if (paramType === 'number' && !isNaN(Number(templated))) {
                    return Number(templated);
                  } else if (paramType === 'boolean') {
                    return templated === 'true' || templated === '1' || templated === 'yes';
                  }
                }
                return templated;
              }
              return null;
            } else if (typeof item === 'object' && item !== null && !Array.isArray(item)) {
              const templated = this.templateObject(
                item as Record<string, unknown>,
                params,
                paramDefinitions
              );
              // Only include if has content
              return Object.keys(templated).length > 0 ? templated : null;
            }
            return item;
          })
          .filter((item) => item !== null);
        // Only include array if it has items
        if (templatedArray.length > 0) {
          result[key] = templatedArray;
        }
      } else if (value !== undefined && value !== null) {
        result[key] = value;
      }
    }
    return result;
  }
}

export default HttpExecutor;

File: python/packages/core/src/matimo/executors/http_executor.py
"""
HTTP executor — makes HTTP requests for 'type: http' tools.
Mirrors: packages/core/src/executors/http-executor.ts
"""
from __future__ import annotations

import base64
import logging
import re
import time
from typing import Any

import httpx

from matimo.core.models import HttpExecution, ToolDefinition
from matimo.encodings.parameter_encoding import apply_parameter_encodings
from matimo.errors import ErrorCode, MatimoError, from_http_error

logger = logging.getLogger("matimo")

_PLACEHOLDER_RE = re.compile(r"\{([^}]+)\}")


class HttpExecutor:
    """
    Executes HTTP tools using httpx.
    Mirrors: HttpExecutor in http-executor.ts

    Security:
    - Credentials are never logged.
    - URL placeholders are validated before request dispatch.
    - Basic auth credentials are base64-encoded per RFC 7617.
    """

    async def execute(
        self,
        tool: ToolDefinition,
        params: dict[str, Any],
        credentials: dict[str, str] | None = None,
    ) -> Any:  # noqa: ANN401
        """
        Execute an HTTP tool.

        Returns Any because tools return parsed JSON (dict/list/str/number) or raw text.

        Raises:
            MatimoError on validation, timeout, or HTTP error.
        """
        exec_cfg = tool.execution
        if not isinstance(exec_cfg, HttpExecution):
            raise MatimoError(
                f"Tool '{tool.name}' is not an HTTP tool",
                ErrorCode.EXECUTION_FAILED,
                {"tool_name": tool.name, "execution_type": exec_cfg.type},
            )

        start = time.monotonic()

        # 1. Apply parameter encodings (e.g. Gmail MIME base64url)
        working_params = dict(params)
        if exec_cfg.parameter_encoding:
            encodings = [e.model_dump() for e in exec_cfg.parameter_encoding]
            working_params = apply_parameter_encodings(working_params, encodings)

        # 2. Build credential lookup: per-call overrides take priority
        creds = dict(credentials or {})

        # 3. Template URL + validate all placeholders are present
        url = self._template(exec_cfg.url, working_params, creds)

        # 4. Template headers
        headers: dict[str, str] = {}
        for k, v in (exec_cfg.headers or {}).items():
            headers[k] = self._template(v, working_params, creds)

        # 5. Handle Basic auth injection
        if tool.authentication and tool.authentication.type and tool.authentication.type.value == "basic":
            headers = self._apply_basic_auth(headers, tool, creds)

        # 6. Build query params — skip optional params that were not provided
        query_params: dict[str, str] = {}
        for k, v in (exec_cfg.query_params or exec_cfg.params or {}).items():
            # If the value is a bare placeholder like "{cursor}", check whether the
            # referenced param is optional and was not supplied — if so, omit it.
            bare_match = re.fullmatch(r"\{([^}]+)\}", str(v)) if v is not None else None
            if bare_match:
                placeholder_name = bare_match.group(1)
                param_def = (tool.parameters or {}).get(placeholder_name)
                is_optional = param_def is not None and not getattr(param_def, "required", True)
                if is_optional and placeholder_name not in (working_params or {}):
                    continue
            try:
                query_params[k] = self._template(v, working_params, creds)
            except MatimoError as exc:
                if exc.code == ErrorCode.INVALID_PARAMETER:
                    # Optional param not provided — omit from query string
                    param_def = (tool.parameters or {}).get(k)
                    is_optional = param_def is None or not getattr(param_def, "required", True)
                    if is_optional:
                        continue
                raise

        # 7. Template body
        body = self._template_object(exec_cfg.body, working_params, creds, tool.parameters)

        # 8. Determine Content-Type and prepare request body
        content_type = headers.get("Content-Type", headers.get("content-type", ""))
        request_content: Any = None
        request_data: Any = None
        request_json: Any = None

        if body is not None:
            if "application/x-www-form-urlencoded" in content_type:
                request_data = body
            else:
                request_json = body

        # 9. Dispatch
        timeout_ms = exec_cfg.timeout or 30_000
        timeout_s = timeout_ms / 1000.0

        try:
            async with httpx.AsyncClient(timeout=timeout_s) as client:
                response = await client.request(
                    method=exec_cfg.method,
                    url=url,
                    headers=headers,
                    params=query_params if query_params else None,
                    json=request_json,
                    data=request_data,
                    content=request_content,
                )
                response.raise_for_status()
        except httpx.TimeoutException as exc:
            duration = time.monotonic() - start
            raise MatimoError(
                f"Tool '{tool.name}' timed out after {duration:.1f}s",
                ErrorCode.TIMEOUT,
                {"tool_name": tool.name, "timeout_ms": timeout_ms},
                cause=exc,
            ) from exc
        except httpx.HTTPStatusError as exc:
            raise from_http_error(exc, f"HTTP error executing tool '{tool.name}'") from exc
        except httpx.RequestError as exc:
            raise MatimoError(
                f"Network error executing tool '{tool.name}': {exc}",
                ErrorCode.NETWORK_ERROR,
                {"tool_name": tool.name},
                cause=exc,
            ) from exc

        # 10. Parse response
        content_type_resp = response.headers.get("content-type", "")
        if "application/json" in content_type_resp:
            return response.json()
        return response.text

    # ------------------------------------------------------------------
    # Templating helpers
    # ------------------------------------------------------------------

    def _template(
        self,
        value: str,
        params: dict[str, Any],
        creds: dict[str, str],
    ) -> str:
        """
        Replace {placeholder} tokens in a string with parameter or credential values.
        Raises MatimoError(INVALID_PARAMETER) if a placeholder has no value.
        """
        def replacer(m: re.Match[str]) -> str:
            name = m.group(1)
            # Params take priority, then env-level credentials
            if name in params:
                return str(params[name])
            if name in creds:
                return creds[name]
            # Fall back to env var of the same name (single-tenant mode)
            import os
            env_val = os.environ.get(name)
            if env_val is not None:
                return env_val
            raise MatimoError(
                f"Missing value for placeholder '{{{name}}}'",
                ErrorCode.INVALID_PARAMETER,
                {"placeholder": name},
            )

        return _PLACEHOLDER_RE.sub(replacer, value)

    def _template_object(
        self,
        obj: object,
        params: dict[str, Any],
        creds: dict[str, str],
        param_definitions: dict[str, Any] | None = None,
    ) -> Any:  # noqa: ANN401
        # Returns the same JSON-like structure (str/dict/list/None) — Any is accurate.
        """Recursively template strings within dicts/lists.

        When a string value is exactly a single placeholder (e.g. ``"{page_size}"``)
        and the corresponding parameter is typed ``number`` or ``boolean`` in the tool
        schema, the substituted value is coerced to the correct Python type so that
        JSON serialisation sends the right type (e.g. ``5`` not ``"5"``).
        Also mirrors the TypeScript behaviour of embedding ``object``/``array``
        parameters directly without stringification.
        """
        if obj is None:
            return None
        if isinstance(obj, str):
            single_match = re.fullmatch(r"\{([a-zA-Z_][a-zA-Z0-9_]*)\}", obj)
            if single_match:
                param_name = single_match.group(1)
                param_value = params.get(param_name)
                # Embed object/array parameters directly (no stringification)
                if param_value is not None and isinstance(param_value, (dict, list)):
                    return param_value
                # Coerce to correct primitive type based on schema
                if param_value is not None and param_definitions:
                    param_def = param_definitions.get(param_name)
                    if param_def is not None:
                        param_type = getattr(param_def, "type", None)
                        if param_type == "number":
                            try:
                                float_val = float(param_value)
                                if int(float_val) == float_val:
                                    return int(param_value)
                                return float_val
                            except (ValueError, TypeError):
                                pass
                        elif param_type == "boolean":
                            return param_value in (True, "true", "1", "yes")
            templated = self._template(obj, params, creds)
            return templated
        if isinstance(obj, dict):
            result: dict[str, Any] = {}
            for k, v in obj.items():
                try:
                    templated = self._template_object(v, params, creds, param_definitions)
                    # Skip None and empty nested dicts (mirrors TS: "Only include nested object if it has content")
                    if isinstance(templated, dict) and not templated:
                        continue
                    if templated is None:
                        continue
                    result[k] = templated
                except MatimoError as exc:
                    if exc.code == ErrorCode.INVALID_PARAMETER:
                        param_def = (param_definitions or {}).get(k)
                        is_optional = param_def is None or not getattr(param_def, "required", True)
                        if is_optional:
                            continue
                    raise
            return result
        if isinstance(obj, list):
            return [self._template_object(item, params, creds, param_definitions) for item in obj]
        return obj

    # ------------------------------------------------------------------
    # Auth helpers
    # ------------------------------------------------------------------

    def _apply_basic_auth(
        self,
        headers: dict[str, str],
        tool: ToolDefinition,
        creds: dict[str, str],
    ) -> dict[str, str]:
        """
        Inject HTTP Basic authentication header.
        Reads username/password from per-call credentials or process env.
        SECURITY: credentials are never logged.
        """
        import os

        auth_cfg = tool.authentication
        if auth_cfg is None:
            return headers

        username_env = auth_cfg.username_env or f"{tool.name.upper()}_USERNAME"
        password_env = auth_cfg.password_env or f"{tool.name.upper()}_PASSWORD"

        username = creds.get(username_env) or os.environ.get(username_env, "")
        password = creds.get(password_env) or os.environ.get(password_env, "")

        if not username and not password:
            return headers  # no basic creds available — let request proceed

        token = base64.b64encode(f"{username}:{password}".encode()).decode()
        headers = dict(headers)
        headers["Authorization"] = f"Basic {token}"
        return headers
